plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+
+	/* from device */
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					    G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    HID_REPORT_GET,
+					    (HID_REPORT_TYPE_INPUT << 8) | ReportID,
+					    FU_CXAUDIO_HID_INTERFACE,
+					    buf, bufsz, &actual_length,
+					    FU_CXAUDIO_USB_TIMEOUT, NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_CXAUDIO_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "HID::READ", buf, bufsz);
+	if (bufsz != actual_length) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "read 0x%x bytes of expected 0x%x",
+			     (guint) actual_length, bufsz);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+typedef enum {
+	FU_CXAUDIO_OPERATION_READ,
+	FU_CXAUDIO_OPERATION_WRITE,
+	FU_CXAUDIO_OPERATION_LAST
+} FuCxaudioOperation;
+
+typedef enum {
+	FU_CXAUDIO_OPERATION_FLAG_NONE			= 0,
+	FU_CXAUDIO_OPERATION_FLAG_VERIFY		= (1 << 4),
+} FuCxaudioOperationFlags;
+
+static gboolean
+fu_cxaudio_device_operation (FuCxaudioDevice *self,
+			     FuCxaudioOperation operation,
+			     FuCxaudioMemKind mem_kind,
+			     guint32 addr,
+			     guint8 *buf,
+			     guint32 bufsz,
+			     FuCxaudioOperationFlags flags,
+			     GError **error)
+{
+	const guint32 idx_read = 0x1;
+	const guint32 idx_write = 0x5;
+	const guint32 payload_max = 0x20;
+	guint32 size = 0x02800;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	g_return_val_if_fail (bufsz > 0, FALSE);
+	g_return_val_if_fail (buf != NULL, FALSE);
+
+	/* check if memory operation is supported by device */
+	if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+	    mem_kind == FU_CXAUDIO_MEM_KIND_CPX_ROM) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "trying to write unwritable section %u",
+			     mem_kind);
+		return FALSE;
+	}
+
+	/* check memory address - should be within valid range */
+	switch (mem_kind) {
+	case FU_CXAUDIO_MEM_KIND_EEPROM:
+		size = 0x20000;
+		break;
+	default:
+		break;
+	}
+	if (addr > size) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "address out of range 0x%x < 0x%x",
+			     addr, size);
+		return FALSE;
+	}
+
+	/* send to hardware */
+	chunks = fu_chunk_array_new (buf, bufsz, addr, 0x0, payload_max);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chunk = g_ptr_array_index (chunks, i);
+		guint8 inbuf[FU_CXAUDIO_INPUT_REPORT_SIZE] = { 0 };
+		guint8 outbuf[FU_CXAUDIO_OUTPUT_REPORT_SIZE] = { 0 };
+
+		/* first byte is always report ID */
+		outbuf[0] = FU_CXAUDIO_MEM_WRITEID;
+
+		/* set memory address and payload length (if relevant) */
+		if (chunk->address >= 64 * 1024)
+			outbuf[1] |= 1 << 4;
+		outbuf[2] = chunk->data_sz;
+		fu_common_write_uint16 (outbuf + 3, chunk->address, G_BIG_ENDIAN);
+
+		/* set the write bit */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE)
+			outbuf[1] |= 1 << 6;
```

Comment:
I think you can bring this into the if block below, the order won't matter before or after setting mem type.

Reply:
Agreed, done.